src/placeos-driver/interface/sensor.cr

```diff
@@ -0,0 +1,137 @@
+require "json"
+require "s2_cells"
+
+module PlaceOS::Driver::Interface::Sensor
+  enum SensorType
+    Temperature # https://en.wikipedia.org/wiki/Celsius
+    Humidity    # https://en.wikipedia.org/wiki/Humidity#Relative_humidity (percentage)
+    Illuminance # https://en.wikipedia.org/wiki/Illuminance (lux)
+    Pressure    # https://en.wikipedia.org/wiki/Pascal_(unit)
+    Trigger     # Proximity or motion etc
+    Switch      # On or Off - water leak, door open etc
+    Level       # Percentage + raw value + technical human readable (battery, fuel, water tank, oxygen, Co2)
+    Flow        # https://en.wikipedia.org/wiki/Cubic_metre_per_second (gas or liquid)
+    Counter
+    Acceleration # MetrePerSecondSquared
+    Speed        # MetrePerSecond
+    # Gyroscope: roll, pitch, yaw - Angle
+    Roll
+    Pitch
+    Yaw
+    Compass              # direction (magnetic field)
+    Current              # https://en.wikipedia.org/wiki/Ampere
+    Voltage              # https://en.wikipedia.org/wiki/Volt
+    ElectricalResistance # https://en.wikipedia.org/wiki/Ohm
+    Radiation            # https://en.wikipedia.org/wiki/Sievert
+    Distance
+  end
+
+  # Using SI units and SI derived units
+  enum Unit
+    Celsius               # Temperature - celsius over Kelvin to avoid some conversions
+    Percentage            # Humidity, Level
+    Lux                   # Illuminance
+    Pascal                # Pressure
+    Boolean               # Trigger, Switch
+    Integer               # Counter
+    CubicMetrePerSecond   # Liquid or gas flow rate
+    MetrePerSecondSquared # Acceleration
+    MetrePerSecond        # Speed
+    Angle                 # Compass, Accel, Gyro
+    Ampere                # Current
+    Volt                  # Voltage
+    Ohm                   # ElectricalResistance
+    Sievert               # Radiation
+    Metre                 # Distance
+  end
+
+  # return the specified sensor details
+  abstract def sensor(unique_id : String) : Detail?
+
+  # return an array of sensor details
+  # zone_id can be ignored if location is unknown by the sensor provider
+  # mac_address can be used to grab data from a single device (basic grouping)
+  abstract def sensors(type : String? = nil, mac_address : String? = nil, zone_id : String? = nil) : Array(Detail)
+
+  abstract class Location
+    include JSON::Serializable
+
+    use_json_discriminator "type", {
+      "geo" => GeoLocation,
+      "map" => MapLocation,
+    }
+  end
+
+  class GeoLocation < Location
+    getter type : String = "geo"
+
+    def initialize(@lat, @lon)
+    end
+
+    property lat : Float64
+    property lon : Float64
+
+    def s2_cell_id(s2_level : Int32 = 21)
+      S2Cells::LatLon.new(lat, lon).to_token(s2_level)
+    end
+  end
+
+  class MapLocation < Location
+    getter type : String = "map"
+
+    def initialize(@x, @y)
```

Comment:
Are these expected to be `0..1`? If so, might be good to enforce here.

Reply:
nah, this is raw data - such as that coming out of Cisco DNA Spaces
Area Management will convert this into a percentage